src/conn/api_calc_modify.c

```diff
@@ -0,0 +1,193 @@
+/*-
+ * Copyright (c) 2014-2018 MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+#define	WT_CM_BLOCKSIZE 8
+#define	WT_CM_MINMATCH 64
+#define	WT_CM_STARTGAP (WT_CM_BLOCKSIZE / 2)
+
+typedef struct {
+	const uint8_t *s1, *e1; /* Start / end of pre-image. */
+	const uint8_t *s2, *e2; /* Start / end of after-image. */
+
+	size_t maxdiff;
+	int maxentries;
+} WT_CM_STATE;
+
+typedef struct {
+	const uint8_t *m1, *m2;
+	size_t len;
+} WT_CM_MATCH;
+
+/*
+ * __cm_extend --
+ *	Given a potential match size, extend to find the complete match.
+ */
+static void
+__cm_extend(WT_CM_STATE *cms,
+    const uint8_t *m1, const uint8_t *m2, WT_CM_MATCH *match)
+{
+	const uint8_t *p1, *p2;
+
+	/* Step past the end and before the beginning of the matching block. */
+	for (p1 = m1, p2 = m2;
+	    p1 < cms->e1 && p2 < cms->e2 && *p1 == *p2;
+	    p1++, p2++)
+		;
+
+	for (; m1 >= cms->s1 && m2 >= cms->s2 && *m1 == *m2;
+	    m1--, m2--)
+		;
+
+	match->m1 = m1 + 1;
+	match->m2 = m2 + 1;
+	match->len = p1 > m1 ? (size_t)(p1 - m1 - 1) : 0;
+}
+
+/*
+ * __cm_hash --
+ *	Calculate a "hash" of a block of bytes.
+ */
+static uint64_t
+__cm_hash(const uint8_t *p)
+{
+	uint64_t h;
+
+	memcpy(&h, p, sizeof h);
+#ifndef WORDS_BIGENDIAN
+	h = __wt_bswap64(h);
+#endif
+	return (h);
+}
+
+/*
+ * __cm_add_modify --
+ *	Add a modify operation to the list of entries.
+ *
+ *	Fails if all entries are used or the maximum bytes of difference is
+ *	exceeded.
+ */
+static int
+__cm_add_modify(WT_CM_STATE *cms, const uint8_t *p1, size_t len1,
+    const uint8_t *p2, size_t len2, WT_MODIFY *entries, int *nentriesp)
+{
+	WT_MODIFY *mod;
+
+	if (*nentriesp >= cms->maxentries || len2 > cms->maxdiff)
+		return (WT_NOTFOUND);
+
+	mod = entries + (*nentriesp)++;
+	mod->size = len1;
+	mod->offset = (size_t)(p1 - cms->s1);
+	mod->data.data = p2;
+	mod->data.size = len2;
+	cms->maxdiff -= len2;
+
+	return (0);
+}
+
+/*
+ * wiredtiger_calc_modify --
+ *	Calculate a set of WT_MODIFY operations to represent an update.
+ */
+int
+wiredtiger_calc_modify(const WT_ITEM *prev, const WT_ITEM *newv,
+    size_t maxdiff, WT_MODIFY *entries, int *nentriesp)
+{
+	WT_CM_MATCH match;
+	WT_CM_STATE cms;
+	size_t gap, i;
+	uint64_t h, hend, hstart;
+	const uint8_t *p1, *p2, *used1, *used2;
+	int size_delta;
```

Comment:
Shouldn't `size_delta` be a `size_t`, it's accumulating memory sizes?

Reply:
Yes, except it's signed (bytes can be added or removed), so `ssize_t`...